Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp

```diff
@@ -0,0 +1,687 @@
+// Copyright 2022 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "Core/HW/EXI/BBA/BuiltIn.h"
+
+#include <cstring>
+
+#include <SFML/Network.hpp>
+
+#include "Common/Assert.h"
+#include "Common/Logging/Log.h"
+#include "Common/MsgHandler.h"
+#include "Common/StringUtil.h"
+#include "Common/Swap.h"
+#include "Core/HW/EXI/EXI_Device.h"
+#include "Core/HW/EXI/EXI_DeviceEthernet.h"
+#include "VideoCommon/OnScreenDisplay.h"
+
+// #define BBA_TRACK_PAGE_PTRS
+
+namespace ExpansionInterface
+{
+u64 GetTickCountStd()
+{
+  using namespace std::chrono;
+  return duration_cast<milliseconds>(steady_clock::now().time_since_epoch()).count();
+}
+
+bool CEXIETHERNET::BuiltInBBAInterface::Activate()
+{
+  if (IsActivated())
+    return true;
+
+  active = true;
+  m_in_frame = std::make_unique<u8[]>(9004);
+  m_out_frame = std::make_unique<u8[]>(9004);
+  for (auto& buf : queue_data)
+    buf = std::make_unique<u8[]>(2048);
+  fake_mac = Common::GenerateMacAddress(Common::MACConsumer::BBA);
+  // clear all ref
+  for (int i = 0; i < std::size(network_ref); i++)
+  {
+    network_ref[i].ip = 0;
+  }
+
+  return RecvInit();
+}
+
+void CEXIETHERNET::BuiltInBBAInterface::Deactivate()
+{
+  // Is the BBA Active? If not skip shutdown
+  if (!IsActivated())
+    return;
+  // Signal read thread to exit.
+  m_read_enabled.Clear();
+  m_read_thread_shutdown.Set();
+  active = false;
+
+  // kill all active socket
+  for (int i = 0; i < std::size(network_ref); i++)
+  {
+    if (network_ref[i].ip != 0)
+    {
+      network_ref[i].type == IPPROTO_TCP ? network_ref[i].tcp_socket.disconnect() :
+                                           network_ref[i].udp_socket.unbind();
+    }
+    network_ref[i].ip = 0;
+  }
+
+  // Wait for read thread to exit.
+  if (m_read_thread.joinable())
+    m_read_thread.join();
+}
+
+bool CEXIETHERNET::BuiltInBBAInterface::IsActivated()
+{
+  return active;
+}
+
+void CEXIETHERNET::BuiltInBBAInterface::WriteToQueue(u8* data, int length)
+{
+  queue_data_size[queue_write] = length;
+  memcpy(&queue_data[queue_write][0], data, length);
+  if (((queue_write + 1) & 15) == queue_read)
+  {
+    return;
+  }
+  queue_write = (queue_write + 1) & 15;
+}
+
+void CEXIETHERNET::BuiltInBBAInterface::HandleARP(Common::EthernetHeader* hwdata,
+                                                  Common::ARPHeader* arpdata)
+{
+  memset(&m_in_frame[0], 0, 0x30);
+  Common::EthernetHeader* hwpart = (Common::EthernetHeader*)&m_in_frame[0];
+  *hwpart = Common::EthernetHeader(ARP_PROTOCOL);
+  memcpy(&hwpart->destination, &m_eth_ref->mBbaMem[BBA_NAFR_PAR0], Common::MAC_ADDRESS_SIZE);
+  hwpart->source = fake_mac;
+
+  Common::ARPHeader* arppart = (Common::ARPHeader*)&m_in_frame[14];
+  Common::MACAddress bba_mac;
+
+  memcpy(&bba_mac, &m_eth_ref->mBbaMem[BBA_NAFR_PAR0], Common::MAC_ADDRESS_SIZE);
+  if (arpdata->target_ip == 0xc701a8c0)
```

Comment:
A constant should be defined for this value `0xc701a8c0` as it's reused at several places.